### PR TITLE
Make more functions tail-recursive, and use tail-rec List functions

### PR DESCRIPTION
**Pull request by gpetiot**
The majority of changes are straightforward, but the commit 826915d could use a review (not sure if it really improves the situation and we lost in readability).

### CODE REVIEW

File: src/lib/uTop.ml [hunk 1]
```diff
@@ -104,7 +105,8 @@ let default_keywords = [
   "when"; "while"; "with"; "try_lwt"; "finally"; "for_lwt"; "lwt";
 ]
 
-let keywords = ref (List.fold_right String_set.add default_keywords String_set.empty)
+let add set elt = String_set.add elt set
+let keywords = ref (List.fold_left add String_set.empty default_keywords)
```
Comment (diml) [inline src/lib/uTop.ml]: What about using `String_set.of_list`?

File: src/lib/uTop_history.ml [hunk 2]
```diff
@@ -33,15 +35,18 @@ let strip_colors s =
...
+      match find_escape offset with
+      | None -> (offset, len) :: acc
+      | Some esc_offset ->
+         try
+           let i = String.index_from s esc_offset 'm' in
+           aux ((offset, esc_offset) :: acc) (i + 1)
+         with
+         | Not_found -> (offset, len) :: acc
+    in
+    List.rev (aux [] offset)
```
Comment (diml) [inline src/lib/uTop_history.ml]: Note that we do a `List.map` on the result, so we might as well not do this `List.rev` and use `List.rev_map`

File: src/lib/uTop_styles.ml [hunk 3]
```diff
@@ -98,87 +98,107 @@ let load () =
         Lwt_log.error_f "cannot load styles from %S: %s: %s" fn func (Unix.error_message error)
     | exn -> Lwt.fail exn)
 
-let rec stylise_filter_layout stylise tokens =
-  match tokens with
+let stylise_filter_layout stylise tokens =
```
Comment (diml) [inline src/lib/uTop_styles.ml]: We can probably make this code clearer by using something like `List.filter_map`

File: src/top/expunge/jbuild [hunk 4]
```diff
@@ -1,5 +1,7 @@
 (jbuild_version 1)
 
+(rule (copy ../../lib/uTop_utils.ml uTop_utils.ml))
```
Comment (diml) [inline src/top/expunge/jbuild]: Why not simply use the utop library in expunge?

**Comment by diml** [comment]
I'm not sure about 826915d either: the input of the `stylise` functions comes from the user, so I think we'd overflow the terminal buffer way before we blow up utop's stack. In the end it's better to keep the code simple.

File: src/lib/uTop_utils.ml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  let map2 f l l' =
+    List.rev (List.rev_map2 f l l')
+
+  let concat l =
+    let rec aux acc = function
+    | [] -> acc
+    | h :: t ->
+       aux (List.rev_append h acc) t
+    in
+    List.rev (aux [] l)
```
Comment (diml) [inline src/lib/uTop_utils.ml]: I'm wondering if we shouldn't just use a stdlib more systematically for community projects. That would make maintaining them easier overall

**Comment by diml** [comment]
That's fine, however we'd have to drop compatibility with OCaml < 4.04

**Comment by rgrinberg** [comment]
That's acceptable. Utop isn't an essential project like dune that keeping strict compatibility with ancient versions of OCaml is necessary.

**Comment by pmetzger** [comment]
If we adopt base, the possibility of ultimately merging this into the main OCaml release goes away, so that might also be a consideration.

**Comment by pmetzger** [comment]
(It would be nice to replace the default toplevel with utop, though I know that's a bit of a dream.)

**Comment by rgrinberg** [comment]
Yeah, I don't think it's desirable to intertwine utop's development with the rest of OCaml anyways.

**Comment by diml** [comment]
So far, we have been trying to take things out of the compiler, so it seems unlikely that utop would go in. There is some value in trying to keep the dependencies of utop to a minimum: it's easier to get it up and running with development compilers. However, there is a maintenance cost as well.

In general, using a stdlib for community maintained projects seems like a win to me as it avoids maintaining the same utility functions in many places.

**Comment by pmetzger** [comment]
> In general, using a stdlib for community maintained projects seems like a win to me as it avoids maintaining the same utility functions in many places.

Sure, but what's missing from the default stdlib that utop needs at this point?

**Comment by diml** [comment]
This PR adds a `UTop_utils` module that looks like a mini-stdlib. Looking at it again, it simply redefines a couple of functions from the stdlib to be tail-rec, so it doesn't seem worth having it in utop. 

@gpetiot could you remove `UTop_utils` from this PR? There is nothing in utop that would justify not using the standard versions.

File: src/lib/uTop_styles.ml [hunk 6]
```diff
@@ -98,21 +98,20 @@ let load () =
         Lwt_log.error_f "cannot load styles from %S: %s: %s" fn func (Unix.error_message error)
     | exn -> Lwt.fail exn)
 
-let rec stylise_filter_layout stylise tokens =
-  match tokens with
-    | [] ->
-        []
-    | (Comment (Comment_reg, _), loc) :: tokens ->
+let stylise_filter_layout stylise tokens =
+  let rec aux acc = function
```
Comment (diml) [inline src/lib/uTop_styles.ml]: This function doesn't look recursive anymore

**Comment by diml** [comment]
Merged, thanks!

**Pull request merged**